### PR TITLE
[FIXED JENKINS-33425] Refactored the cleanUp() method to log progress and push for completion

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-33425](https://issues.jenkins-ci.org/browse/JENKINS-33425)

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -2823,73 +2914,313 @@ public void cleanUp() {
...
+            } catch (Throwable e) {
+                LOGGER.log(SEVERE, "Failed to stop plugin manager", e);
+                // save for later
+                errors.add(e);
+            }
+        }
+    }
 
-        if(getRootDir().exists())
+    private void _cleanUpPersistQueue(List<Throwable> errors) {
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: :+1: Nice, no more loss of the build queue during shutdown.

Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: Well if you do a `System.exit(0)` or `kill -9` then there is not much we can do... but yes, normal shutdown should keep the queue

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -2812,9 +2818,94 @@ public synchronized void save() throws IOException {
...
+                boolean first = true;
+                for (Throwable e : errors) {
+                    if (first) {
+                        first = false;
+                    } else {
+                        message.append("; ");
+                    }
+                    message.append(e.getMessage());
+                }
+                throw new RuntimeException(message.toString(), errors.get(0));
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Could use `Throwable.addSuppressed` on the others.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 3]
```diff
@@ -2823,73 +2914,313 @@ public void cleanUp() {
                 public void execute(Runnable command) {
                     command.run();
                 }
+            }, new ReactorListener() {
+                final Level level = Level.parse(Configuration.getStringConfigParameter("termLogLevel", "FINE"));
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: :ant: Why should this be a configuration option? Log at `FINE` and if the user wishes to see otherwise hidden messages, they can create a custom logger for that purpose.

Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: symmetry with `initLogLevel`

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -2823,73 +2914,313 @@ public void cleanUp() {
...
+                    } catch (OutOfMemoryError e) {
+                        // we should just propagate this, no point trying to log
+                        throw e;
+                    } catch (LinkageError e) {
+                        LOGGER.log(Level.WARNING, "Could not disconnect " + c + ": " + e.getMessage(), e);
+                        // safe to ignore and continue for this one
+                    } catch (Throwable e) {
+                        LOGGER.log(Level.WARNING, "Could not disconnect " + c + ": " + e.getMessage(), e);
+                        // save for later
+                        errors.add(e);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: :ant: Yikes, this is a lot of duplication. Why not introduce a helper method taking `Callable<Void>` (lambdas, eventually) and handling the boilerplate of catching and reporting exceptions?

**Comment by jglick** [comment]
Would be a :bee: except for the merge conflicts.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 2 above)
Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: Thought that was Java 8... but phew it's Java 7

**Comment by stephenc** [comment]
@jglick you should be all good now ;-)

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 4 above)
Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: I had to tune quite a few of these based on expected exceptions, I had started off with a `Runnable` wrapper, but IMHO that got messy for the specific cases so I just inlined it and fixed by hand

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 5]
```diff
@@ -2815,9 +2821,99 @@ public synchronized void save() throws IOException {
...
+    private void fireBeforeShutdown(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Notifying termination");
+        for (ItemListener l : ItemListener.all()) {
+            try {
+                l.onBeforeShutdown();
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(Level.WARNING, "ItemListener " + l + ": " + e.getMessage(), e);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: I was sure that the message from the exception gets printed to the logs - so this would duplicate that.

Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: You would be surprised how many people fail to read the exception stack trace correctly

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 6]
```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
...
+                LOGGER.log(SEVERE, "Failed to shutdown UDP Broadcast Thread", e);
+                // save for later
+                errors.add(e);
+            }
+        }
+    }
+
+    private void _cleanUpCloseDNSMulticast(List<Throwable> errors) {
+        if(dnsMultiCast!=null) {
+            LOGGER.log(Level.FINE, "Closing DNS multi-cast service-discovery");
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: :ant: multicast is one word.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 7]
```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
...
+            try {
+                dnsMultiCast.close();
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(SEVERE, "Failed to close DNS Multicast service", e);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: see - no hyphen here :-)

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 8]
```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
...
-        if (timer != null) {
-            timer.cancel();
+    private void _cleanUpShutdownTriggers(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Shutting down triggers");
+        try {
+            java.util.Timer timer = Trigger.timer;
+            if (timer != null) {
+                timer.cancel();
+            }
+            // TODO: how to wait for the completion of the last job?
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: you schedule a task on the timer to run immediately and that task cancels the timer.  This guarantees it is the last task to execute.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 9]
```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
...
+        } catch (Throwable e) {
+            LOGGER.log(SEVERE, "Failed to shut down triggers", e);
+            // save for later
+            errors.add(e);
         }
-        // TODO: how to wait for the completion of the last job?
-        Trigger.timer = null;
+    }
 
-        Timer.shutdown();
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: why no longer shuting down the `Timer`'s `ExecutorService`?

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 6 above)
Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/DNSMultiCast.java#L69

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 10]
```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
...
+                errors.add(e);
+            }
+        }
+    }
 
-        if(pluginManager!=null) // be defensive. there could be some ugly timing related issues
-            pluginManager.stop();
+    private void _cleanUpShutdownPluginManager(List<Throwable> errors) {
+        if(pluginManager!=null) {// be defensive. there could be some ugly timing related issues
+            LOGGER.log(Level.INFO, "Stopping plugin manager");
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: why is this log entry INFO yet the other shutting down XXX are FINE?

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 11]
```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
...
+        }
+    }
 
-        theInstance = null;
-        if (JenkinsJVM.isJenkinsJVM()) {
-            JenkinsJVMAccess._setJenkinsJVM(oldJenkinsJVM);
+    private void _cleanUpReleaseAllLoggers(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Releasing all loggers");
+        try {
+            LogFactory.releaseAll();
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: Not part of this change - but this seems wrong as we don't ship with commons-logging but use the SLF4J bridge which uses jul.

**Comment by jtnord** [comment]
:bee: but some comments about logging level inconsistency inline.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 10 above)
Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: This was one that I felt should be user visible

**Comment by stephenc** [comment]
@jtnord @jglick a re-review?

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 12]
```diff
@@ -2815,9 +2823,99 @@ public synchronized void save() throws IOException {
...
+                Iterator<Throwable> iterator = errors.iterator();
+                message.append(iterator.next().getMessage());
+                while (iterator.hasNext()) {
+                    message.append("; ");
+                    message.append(iterator.next().getMessage());
+                }
+                iterator = errors.iterator();
+                RuntimeException exception = new RuntimeException(message.toString(), iterator.next());
+                while (iterator.hasNext()) {
+                    exception.addSuppressed(iterator.next());
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: I suppose I should point out that the two loops could be collapsed into one.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 13]
```diff
@@ -2826,75 +2924,324 @@ public void cleanUp() {
...
+                final CountDownLatch latch = new CountDownLatch(1);
+                timer.schedule(new TimerTask() {
+                    @Override
+                    public void run() {
+                        timer.cancel();
+                        latch.countDown();
+                    }
+                }, 0);
+                if (latch.await(10, TimeUnit.SECONDS)) {
+                    LOGGER.log(Level.FINE, "Triggers shut down successfully");
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Nice, but note that `Trigger.timer` has long been deprecated; is there any equivalent for `jenkins.util.Timer`?

**Comment by jglick** [comment]
:bee:

**Comment by stephenc** [comment]
Test failure seems unrelated

```
hudson.model.AbstractProjectTest.testConfiguringBlockBuildWhenUpstreamBuildingRoundtrip
Took 1 min 55 sec.
Error Message

Read timed out
Stacktrace

java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:152)
	at java.net.SocketInputStream.read(SocketInputStream.java:122)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:139)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:155)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:284)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:140)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:57)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:261)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:165)
	at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:167)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:272)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:124)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:271)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at com.gargoylesoftware.htmlunit.HttpWebConnection.getResponse(HttpWebConnection.java:179)
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1321)
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponse(WebClient.java:1238)
	at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:346)
	at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:415)
	at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:400)
	at org.jvnet.hudson.test.HudsonTestCase$WebClient.goTo(HudsonTestCase.java:1704)
	at org.jvnet.hudson.test.HudsonTestCase$WebClient.goTo(HudsonTestCase.java:1692)
	at org.jvnet.hudson.test.HudsonTestCase$WebClient.getPage(HudsonTestCase.java:1648)
	at org.jvnet.hudson.test.HudsonTestCase$WebClient$getPage.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:42)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
	at hudson.model.AbstractProjectTest.testConfiguringBlockBuildWhenUpstreamBuildingRoundtrip(AbstractProjectTest.groovy:166)
```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 14]
```diff
@@ -2815,9 +2823,99 @@ public synchronized void save() throws IOException {
...
+                while (iterator.hasNext()) {
+                    message.append("; ");
+                    message.append(iterator.next().getMessage());
+                }
+                iterator = errors.iterator();
+                RuntimeException exception = new RuntimeException(message.toString(), iterator.next());
+                while (iterator.hasNext()) {
+                    exception.addSuppressed(iterator.next());
+                }
+                throw exception;
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java:2889]: This seems wrong. It means shutdown will hang if one of the terminators failed, whether that was important or not.

Comment (jglick) [reply core/src/main/java/jenkins/model/Jenkins.java:2889]: > resulting in tests that fail to clean up correctly remaining hidden and consequently resulting in random test failures

sounds like this would better be activated iff `Main.isUnitTest`.
